node/silkworm/trie/intermediate_hashes.cpp

```diff
@@ -336,25 +386,43 @@ static evmc::bytes32 increment_intermediate_hashes(mdbx::txn& txn, const std::fi
 }
 
 // See Erigon (p *HashPromoter) Promote
-static void changed_accounts(mdbx::txn& txn, BlockNum from, PrefixSet& out) {
-    // TODO[Issue 179] delete TrieStorage for deleted accounts
+static PrefixSet gather_changes(mdbx::txn& txn, BlockNum from) {
     const Bytes starting_key{db::block_key(from + 1)};
 
-    auto change_cursor{db::open_cursor(txn, db::table::kAccountChangeSet)};
-    change_cursor.lower_bound(db::to_slice(starting_key), /*throw_notfound=*/false);
-    db::cursor_for_each(change_cursor, [&out](mdbx::cursor&, mdbx::cursor::move_result& entry) {
+    PrefixSet out;
+
+    auto account_changes{db::open_cursor(txn, db::table::kAccountChangeSet)};
+    account_changes.lower_bound(db::to_slice(starting_key), /*throw_notfound=*/false);
+    db::cursor_for_each(account_changes, [&out](mdbx::cursor&, mdbx::cursor::move_result& entry) {
         const ByteView address{db::from_slice(entry.value).substr(0, kAddressLength)};
         const auto hashed_address{keccak256(address)};
-        out.insert(unpack_nibbles(ByteView{hashed_address.bytes, kHashLength}));
+        out.insert(unpack_nibbles(full_view(hashed_address.bytes)));
+        return true;
+    });
+
+    auto storage_changes{db::open_cursor(txn, db::table::kStorageChangeSet)};
+    storage_changes.lower_bound(db::to_slice(starting_key), /*throw_notfound=*/false);
+    // TODO[Issue 179] what happens if there are no storage changes? Will cursor_for_each start from the beginning?
```

Comment:
Wrt this comment the answer is yes: if lower_bound does not find any match (i.e. mdbx::cursor.eof() == true) then for_each will automatically position the cursor to first entry of the table

Reply:
Hmm. To my mind this behaviour of `cursor_for_each` is confusing and undesirable. But let's revisit this in another PR; I don't want to change too much in this one.

Reply:
Well ... the purpose was to save the burden to write boilerplate code for positioning the cursor (to_first/to_last) on those loops which traverse all the buckets' records.
I believe the syntax to prevent the auto-positioning is nevertheless very easy:
```
   auto storage_changes{db::open_cursor(txn, db::table::kStorageChangeSet)};
   if(storage_changes.lower_bound(db::to_slice(starting_key), /*throw_notfound=*/false)) {
    // invoke for_each here
   }
```

Reply:
Solved by 13855bc